# Урок 2. Анализ датасета с помощью Pandas
 ___

Описание данных:
id - unique identified for a house
dateDate - house was sold
pricePrice - is prediction target
bedroomsNumber - of Bedrooms/House
bathroomsNumber - of bathrooms/bedrooms
sqft_livingsquare - footage of the home
sqft_lotsquare - footage of the lot
floorsTotal - floors (levels) in house
waterfront - House which has a view to a waterfront
view - Has been viewed
condition - How good the condition is ( Overall )
grade - overall grade given to the housing unit, based on King County grading system
sqft_above - square footage of house apart from basement
sqft_basement - square footage of the basement
yr_built - Built Year
yr_renovated - Year when house was renovated
zipcode - zip
lat - Latitude coordinate
long - Longitude coordinate
sqft_living15 - The square footage of interior housing living space for the nearest 15 neighbors
sqft_lot15 - The square footage of the land lots of the nearest 15 neighbors

## Условие 1: 1 задача
___
Скачать файл в уроке
Считать данные с помощью pandas
Вывести на экран первые 5 строк
Посмотреть на описание признаков и на их содержание

In [1]:
import pandas as pd

In [2]:
# Считать данные с помощью pandas
kc_house_data = pd.read_csv('kc_house_data.csv', encoding='utf-8')

In [3]:
# Вывести на экран первые 5 строк
kc_house_data.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [4]:
# Посмотреть на описание признаков и на их содержание
kc_house_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

## Условие 2: 2 задача
___
Проведите первичный анализ данных
Изучите типы данных
Найдите количество пропущенных ячеек в данных
Посчитайте основные статистики по всем признакам и поизучайте их
Пишите выводы

In [5]:
# Посчитайте основные статистики по всем признакам и поизучайте их
kc_house_data.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

### Выводы:
(в анализ включен результат работы метода info() выше)
#### _Изучите типы данных_
В датасете преобладают числовые признаки (15 шт. int64 и 5 float64) и один признак с типом данных "object" (дата).
<br>
#### _Найдите количество пропущенных ячеек в данных_
Пропущенные значения отсутствуют. У нас 21613 записи и для каждого признака 21613 non-null вхождения.
<br>
#### _Посчитайте основные статистики по всем признакам и поизучайте их_
Сказать можно следующее:
- каких-то выбросов или аномальных данных не наблюдается (судя по максимальным и минимальным значениям каждого признака)
Смущают максимальные значения некоторых признаков площади, но пока трудно назвать их аномалиями (мало ли продают замок с огромной территории). Стоит держать в голове, что далее записи с этими признаками следует внимательнее проанализировать.

## Условие 3:
___
Ответьте на несколько вопросов
3.1 В каком диапазоне изменяются стоимости недвижимости?
3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?
3.3 Как много домов с разными этажами в данных?
3.4 Насколько хорошие состояния у домов в данных?
3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [14]:
# 3.1 В каком диапазоне изменяются стоимости недвижимости?
print(f"Цена на недвижимость лежит в диапазонах от {min(kc_house_data['price'])} до {max(kc_house_data['price'])}.\n"
      f"Разница составляет {max(kc_house_data['price']) - min(kc_house_data['price'])}")

Цена на недвижимость лежит в диапазонах от 75000.0 до 7700000.0.
Разница составляет 7625000.0


In [25]:
# 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?
kc_house_data['sqft_living_part'] = round(kc_house_data['sqft_living'] / kc_house_data['sqft_lot'], 4) * 100
kc_house_data['sqft_living_part'].sample(5)

12976    29.05
18583    36.07
6663     59.79
21506     5.59
13264    40.75
Name: sqft_living_part, dtype: float64

In [28]:
print(f"Средняя доля жилой площади от общей площади по всем домам: {round(sum(kc_house_data['sqft_living_part']) / kc_house_data.shape[0], 2)}")

Средняя доля жилой площади от общей площади по всем домам: 32.37


In [32]:
# 3.3 Как много домов с разными этажами в данных?
kc_house_data['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

In [36]:
# 3.4 Насколько хорошие состояния у домов в данных?
kc_house_data['condition'].value_counts().sort_index(ascending=False)

5     1701
4     5679
3    14031
2      172
1       30
Name: condition, dtype: int64

In [37]:
# 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?
print(f"Самый первый дом построен в {min(kc_house_data['yr_built'])} году.\nПоследний дом построен в "
      f"{max(kc_house_data['yr_built'])} году")

Самый первый дом построен в 1900 году.
Последний дом построен в 2015 году


## Условие 4
___
Ответьте на несколько вопросов
4.1 Сколько в среднем стоят дома, у которых 2 спальни?
4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?
4.3 Как много домов коснулся ремонт?
4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [42]:
# 4.1 Сколько в среднем стоят дома, у которых 2 спальни?
two_bedrooms_data = kc_house_data.loc[kc_house_data['bedrooms'] == 2]
two_bedrooms_data.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
2   5631500400  20150225T000000  180000.0         2        1.0          770   
11  9212900260  20140527T000000  468000.0         2        1.0         1160   
18    16000397  20141205T000000  189000.0         2        1.0         1200   
23  8091400200  20140516T000000  252700.0         2        1.5         1070   
31  2426039314  20141201T000000  280000.0         2        1.5         1190   

    sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
2      10000     1.0           0     0  ...         770              0   
11      6000     1.0           0     0  ...         860            300   
18      9850     1.0           0     0  ...        1200              0   
23      9643     1.0           0     0  ...        1070              0   
31      1265     3.0           0     0  ...        1190              0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
2       1933             0    98028  47.7379 -122.233           2720   
11      1942             0    98115  47.6900 -122.292           1330   
18      1921             0    98002  47.3089 -122.210           1060   
23      1985             0    98030  47.3533 -122.166           1220   
31      2005             0    98133  47.7274 -122.357           1390   

    sqft_lot15  sqft_living_part  
2         8062              7.70  
11        6000             19.33  
18        5095             12.18  
23        8386             11.10  
31        1756             94.07  

[5 rows x 22 columns]

In [45]:
print(f"Средняя стоимость домов с двумя спальнями: "
      f"{round(sum(two_bedrooms_data['price']) /two_bedrooms_data.shape[0], 2)}")

Средняя стоимость домов с двумя спальнями: 401372.68


In [48]:
# 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?
round(kc_house_data.loc[kc_house_data['price'] > 600000]['sqft_lot'].mean(), 2)

20442.52

In [51]:
# 4.3 Как много домов коснулся ремонт?
kc_house_data.loc[kc_house_data['yr_renovated'] != 0]['yr_renovated'].count()

914

In [53]:
# либо
len(kc_house_data.loc[kc_house_data['yr_renovated'] != 0])

914

In [54]:
# 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?
sum(kc_house_data.loc[kc_house_data['grade'] > 10]['price']) - sum(kc_house_data.loc[kc_house_data['grade'] < 4]['price'])

841915829.0

## Условие 5
___
Ответьте на несколько вопросов
5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?
5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?
5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [60]:
# 5.1 Выберите дом клиенту
# Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?
len(kc_house_data.loc[kc_house_data['waterfront'] == 1].loc[kc_house_data['bathrooms'] >= 3].loc[kc_house_data['sqft_basement'] > 0])

41

In [69]:
# 5.2 Выберите дом клиенту
# Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?
# что такое "красивый вид из окна"? Признак 'view' говорит о том что дом был просмотрен, а не то что вид красивый. Это условие не учитываю.
client_wish = kc_house_data.loc[(kc_house_data['waterfront'] == 1) & (kc_house_data['yr_built'] >= 1980) & (kc_house_data['condition'] == 5)]
client_wish

id             date      price  bedrooms  bathrooms  \
11257  3761100045  20140618T000000  3000000.0         4       4.25   
14682  7278100515  20140821T000000  1295000.0         2       2.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  sqft_above  \
11257         4850     12445     2.0           1     4  ...        3850   
14682         2910     19449     2.0           1     4  ...        1940   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
11257           1000      1989             0    98034  47.7011 -122.244   
14682            970      1985             0    98177  47.7729 -122.393   

       sqft_living15  sqft_lot15  sqft_living_part  
11257           3350       12210             38.97  
14682           2540       23598             14.96  

[2 rows x 22 columns]

In [70]:
min(client_wish['price']), max(client_wish['price'])

(1295000.0, 3000000.0)

In [72]:
# 5.3 Выберите дом клиенту
# Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?
kc_house_data.loc[(kc_house_data['sqft_basement'] == 0) & (kc_house_data['floors'] == 2) & (kc_house_data['price'] <= 150000)]['grade'].mean()

7.0